embeddings/data/datamodule.py

```diff
@@ -195,3 +222,108 @@ def convert_to_features(
         features["labels"] = example_batch[self.target_field]
 
         return features
+
+
+class SequenceLabelingDataModule(HuggingFaceDataModule):
+    DEFAULT_BATCH_ENCODING_KWARGS = {
+        "padding": True,
+        "truncation": True,
+        "is_split_into_words": True,
+        "return_offsets_mapping": True,
+    }
+    IGNORE_INDEX = -100
+
+    def __init__(
+        self,
+        tokenizer_name_or_path: str,
+        dataset_name: str,
+        text_field: str,
+        target_field: str,
+        train_batch_size: int = 32,
+        eval_batch_size: int = 32,
+        label_all_tokens: bool = False,
+        tokenizer_kwargs: Optional[Dict[str, Any]] = None,
+        batch_encoding_kwargs: Optional[Dict[str, Any]] = None,
+        load_dataset_kwargs: Optional[Dict[str, Any]] = None,
+        **kwargs: Any,
+    ):
+        self.text_field = text_field
+        self.label_all_tokens = label_all_tokens
+        self.batch_encoding_kwargs = (
+            batch_encoding_kwargs if batch_encoding_kwargs else self.DEFAULT_BATCH_ENCODING_KWARGS
+        )
+        self.label_to_id: Dict[str, int] = {}
+        self.id_to_label: Dict[int, str] = {}
+        super().__init__(
+            tokenizer_name_or_path=tokenizer_name_or_path,
+            dataset_name=dataset_name,
+            target_field=target_field,
+            train_batch_size=train_batch_size,
+            eval_batch_size=eval_batch_size,
+            tokenizer_kwargs=tokenizer_kwargs,
+            batch_encoding_kwargs=batch_encoding_kwargs,
+            load_dataset_kwargs=load_dataset_kwargs,
+            **kwargs,
+        )
+
+    def prepare_labels(self) -> None:
+        if isinstance(self.dataset["train"].features[self.target_field].feature, ClassLabel):
+            labels = self.dataset["train"].features[self.target_field].feature.names
+            # No need to convert the labels since they are already ints.
+        else:
+            # Create unique label set from train dataset.
+            labels = sorted(
+                {label for column in self.dataset["train"][self.target_field] for label in column}
+            )
+        self.num_classes = len(labels)
+        for i, label in enumerate(labels):
+            self.label_to_id[label] = i
+            self.id_to_label[i] = label
+
```

Comment:
do you have any examples of datasets that for sequence labeling that do not have a label column in  datasets.Sequence(datasets.features.ClassLabel(...) type so I can work this out? kpwr-ner and nkjp-pos from clarin are both already converted to this type

Reply:
i've looked into our example for flair sequence labeling and we do not handle the case where the label column is not Squence(ClassLabel) type

casting could be done like this:

```
labels = sorted(
                {label for column in self.dataset["train"][self.target_field] for label in column}
            )
            new_features = self.dataset["train"].features.copy()
            new_features['labels'] = Sequence(
                feature=ClassLabel(num_classes=len(labels), names=list(labels)))
            self.dataset = self.dataset.cast(new_features)
```

but I was thinking whether we should handle this in our code for lightning dataset

Reply:
this is sorted out now (dataset is casted again after the labels are converted) and error is raised when the dataset is not Sequence(ClassLabel(...)) as in our flair corpus

Reply:
meh, I agree that users should be responsible for preparing the dataset. The prepared code for casting does not work, it is caused by the fact, that `.cast() ` written by HF does not convert the column. It tries to read column as if it were written in `ClassLabel` i.e. integers.


```from datasets import Dataset, ClassLabel
from datasets import Sequence as DatasetsSequence

my_dict = {"t": [["a", "a", "o"], ["b", "o", "a"], ]}
dataset = Dataset.from_dict(my_dict)


new_features = dataset.features.copy()
new_features["t"] = DatasetsSequence(
        feature=ClassLabel(num_classes=3, names=["a", "b", "o"])
    )
dataset = dataset.cast(new_features)```
output: `pyarrow.lib.ArrowInvalid: Failed to parse string: 'a' as a scalar of type int64`

Because of that, though I preferred the first version or the second with raising the error, the casting can be removed.

Reply:
yes, I am aware of the fact that it only works if the labels are integers, but it is the default way to encode the labels in HF from what I see; now we only cast the labels because they are put in Values when calling the convert_to_features method - we just encode them again as labels as they were before. This way we can use the int2str and str2int function for sequence labeling